pgrx/src/callconv.rs

```diff
@@ -283,3 +284,170 @@ where
+        &self,
+        default: DefaultValue,
+    ) -> NonNull<pg_sys::FuncCallContext> {
+        // Safety: User must supply a valid fcinfo to assume_valid() in order
+        // to construct a FcInfo. If that constraint is maintained, this should
+        // be safe.
+        unsafe {
+            let mut flinfo = NonNull::new((*self.0).flinfo).unwrap();
+            if flinfo.as_ref().fn_extra.is_null() {
+                flinfo.as_mut().fn_extra = PgMemoryContexts::For(flinfo.as_ref().fn_mcxt)
+                    .leak_and_drop_on_delete(default())
+                    as crate::void_mut_ptr;
+            }
+
+            // Safety: can new_unchecked() here because we just initialized it.
+            NonNull::new_unchecked(flinfo.as_mut().fn_extra as *mut pg_sys::FuncCallContext)
+        }
+    }
+
+    #[inline]
+    pub fn srf_is_initialized(&self) -> bool {
+        // Safety: User must supply a valid fcinfo to assume_valid() in order
+        // to construct a FcInfo. If that constraint is maintained, this should
+        // be safe.
+        unsafe { !(*(*self.0).flinfo).fn_extra.is_null() }
+    }
+
+    /// Thin wrapper around [`pg_sys::init_MultiFuncCall`], made necessary
+    /// because this structure's FunctionCallInfo is a private field.
+    ///
+    /// This should initialize `self.0.flinfo.fn_extra`
+    #[inline]
+    pub unsafe fn init_multi_func_call(&mut self) -> &'fcx mut pg_sys::FuncCallContext {
+        unsafe {
+            let fcx: *mut pg_sys::FuncCallContext = pg_sys::init_MultiFuncCall(self.0);
+            debug_assert!(fcx.is_null() == false);
+            &mut *fcx
+        }
+    }
+
+    /// Equivalent to "per_MultiFuncCall" with no FFI cost, and a lifetime
+    /// constraint.
+    ///
+    /// Safety: Assumes `self.0.flinfo.fn_extra` is non-null
+    /// i.e. [`FcInfo::srf_is_initialized()`] would be `true`.
+    pub(crate) unsafe fn deref_fcx(&mut self) -> &'fcx mut pg_sys::FuncCallContext {
+        unsafe {
+            let fcx: *mut pg_sys::FuncCallContext = (*(*self.0).flinfo).fn_extra.cast();
+            debug_assert!(fcx.is_null() == false);
+            &mut *fcx
+        }
+    }
+
+    /// Safety: Assumes `self.0.flinfo.fn_extra` is non-null
+    /// i.e. [`FcInfo::srf_is_initialized()`] would be `true`.
+    #[inline]
+    pub unsafe fn srf_return_next(&mut self) {
+        unsafe {
+            self.deref_fcx().call_cntr += 1;
+            (*((*self.0).resultinfo as *mut pg_sys::ReturnSetInfo)).isDone =
+                pg_sys::ExprDoneCond_ExprMultipleResult;
+        }
+    }
+
+    /// Safety: Assumes `self.0.flinfo.fn_extra` is non-null
+    /// i.e. [`FcInfo::srf_is_initialized()`] would be `true`.
+    #[inline]
+    pub unsafe fn srf_return_done(&mut self) {
+        unsafe {
+            pg_sys::end_MultiFuncCall(self.0, self.deref_fcx());
+            (*((*self.0).resultinfo as *mut pg_sys::ReturnSetInfo)).isDone =
+                pg_sys::ExprDoneCond_ExprEndResult;
```

Comment:
can extract the accessor for the resultinfo here as an unsafe fn

Reply:
Resolved